### PR TITLE
core-clp: Refactor `streaming_archive::reader::File::find_message_matching_query`.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# References

# Description

This PR refactors the implementation of query matching in the `clp` File reader. It has two improvements:
1. Move `Query` level checks out of the subquery traversal loop.
2. Rewrite the subquery traversal loop with shorter indent blocks to improve readability.

# Validation performed

Compressed [hive-24hr](https://zenodo.org/records/7094920) dataset and ran wildcard query "*INFO*applic*1*2.6*2?6*application*". Ensured the results returned from `clg` doesn't change after refactoring.

### CODE REVIEW

File: components/core/src/clp/streaming_archive/reader/File.cpp [hunk 1]
```diff
@@ -257,39 +257,32 @@ SubQuery const* File::find_message_matching_query(Query const& query, Message& m
         // Get number of variables in logtype
         auto const& logtype_dictionary_entry = m_archive_logtype_dict->get_entry(logtype_id);
         auto const num_vars = logtype_dictionary_entry.get_num_variables();
+        auto const vars_ix_end{m_variables_ix + num_vars};
+        if (vars_ix_end > m_num_variables) {
+            // Logtypes not in sync with variables, so stop search
+            return nullptr;
+        }
```
Comment (LinZhihao-723) [inline components/core/src/clp/streaming_archive/reader/File.cpp:264]: would this ever happen in a legal archive?

File: components/core/src/clp/streaming_archive/reader/File.cpp [hunk 2]
```diff
@@ -257,39 +257,32 @@ SubQuery const* File::find_message_matching_query(Query const& query, Message& m
         // Get number of variables in logtype
         auto const& logtype_dictionary_entry = m_archive_logtype_dict->get_entry(logtype_id);
         auto const num_vars = logtype_dictionary_entry.get_num_variables();
+        auto const vars_ix_end{m_variables_ix + num_vars};
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/File.cpp:260]: `vars_end_ix` sounds better

File: components/core/src/clp/streaming_archive/reader/File.cpp (hunk 1 above)
Comment (kirkrodrigues) [reply components/core/src/clp/streaming_archive/reader/File.cpp:264]: This should no longer be possible. It used to be possible when we supported storing encoded files on disk, before moving them into an archive.

File: components/core/src/clp/streaming_archive/reader/File.cpp (hunk 2 above)
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/File.cpp:260]:
```suggestion

        auto const vars_ix_end{m_variables_ix + num_vars};
```

File: components/core/src/clp/streaming_archive/reader/File.cpp [hunk 3]
```diff
@@ -257,39 +257,32 @@ SubQuery const* File::find_message_matching_query(Query const& query, Message& m
...
-
-                        matching_sub_query = sub_query;
-                        break;
-                    }
-                }
+        for (auto const* sub_query : query.get_relevant_sub_queries()) {
+            if (false == sub_query->matches_logtype(logtype_id)) {
+                continue;
+            }
+            msg.clear_vars();
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/File.cpp:274]:
```suggestion

            msg.clear_vars();
```

File: components/core/src/clp/streaming_archive/reader/File.cpp [hunk 4]
```diff
@@ -257,39 +257,32 @@ SubQuery const* File::find_message_matching_query(Query const& query, Message& m
...
+            for (auto vars_ix{m_variables_ix}; vars_ix < vars_ix_end; ++vars_ix) {
+                msg.add_var(m_variables[vars_ix]);
+            }
+            if (false == sub_query->matches_vars(msg.get_vars())) {
+                continue;
             }
+            msg.set_logtype_id(logtype_id);
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/File.cpp:281]:
```suggestion

            msg.set_logtype_id(logtype_id);
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

core-clp: Refactor `streaming_archive::reader::File::find_message_matching_query`.

**Pull request merged**